pkg/cloudprovider/types.go

```diff
@@ -44,6 +43,11 @@ type CloudProvider interface {
 	Validate(context.Context, *v1alpha5.Constraints) *apis.FieldError
 	// Name returns the CloudProvider implementation name.
 	Name() string
+	// ResourcePriorities provides information to bin-packing regarding the sort priority of resources. If CPU is before memory
+	// in this list, the bin-packer will prefer to schedule on nodes with less CPU than memory.  This should always include at
+	// least v1.ResourceMemory and v1.ResourceCPU, however it also is suggested to include any specific resource types
+	// (e.g GPUs) ordered before CPU and memory.
+	ResourcePriorities() []v1.ResourceName
```

Comment:
I'm always nervous about expanding the cloud provider API, especially since i'm not sure this is the right semantic. For example, the resource `pod-eni` doesn't matter in terms of cost, we only need to make sure there is "enough". For GPU, you basically never want a GPU instance if your workload doesn't need one. Even if every other instance type is unavailable.

Reply:
I think there's got to be something to support cloud providers that may have new resource types that we aren't aware of.  Consider {cpu: 4, mem: 8, xyz: 1} and {cpu: 4, mem: 8, abc: 2}  , which instance type should we prefer?

Brandon suggested a comparison func, I was looking at something like this to allow the cloud provider to directly provide the comparison:
```
	// LessThan is used by bin-packing to give control to the CloudProvider over how instance types are
	// compared.  If two instance types can both meet the resource requests for a given demand, bin-packing will
	// prefer the 'lesser' one.
	LessThan(lhs, rhs InstanceType) bool
```

There probably does need to be some method for the cloud provider to let bin packing know about those hard preferences (e.g. don't consider GPU instance types for non-GPU workloads).  The old code was hard-coded for the three supported GPU types that were known.  With this PR, we would request both non-GPU and GPU instance types for any non-GPU workload and let `CreateFleet` give us whatever it prefers.

Maybe a compare with a broader return type to allow expressing the incompatibilities?
```
type InstanceCompareResult byte
const (
InstanceCompareResultIncompatible InstanceCompareResult = iota
InstanceCompareResultLessThan
InstanceCompareResultEqual
InstanceCompareResultGreaterThan
``

Reply:
Actually that doesn't work if you compare instance types, but I think it may if you compare v1.ResourceList. You need to both be able to:
- order instance types
- identify if there is some reason why you would prefer not to use an instance type that can satisfy a pod's resource requests

Reply:
Or maybe something like?

```go
	// LessThan is used by bin-packing to give control to the CloudProvider over how instance types are
	// compared.  If two instance types can both meet the resource requests for a given demand, bin-packing will
	// prefer the 'lesser' one.
	LessThan(lhs, rhs InstanceType) bool

	// IsCompatible allows the cloud provider to prevent scheduling a pod against a specific instance type
        // even if it would otherwise normally be possible.  For example, this can be used to prevent pods that
        // don't require GPUs from scheduling against instances that have GPUs.
        IsCompatible(podResources v1.ResourceList, it InstanceType) bool

```

Reply:
This is getting into sketchy territory. It's really increasing the onus of the cloud provider to make decisions about this. In some sense, this really does come down to cost. I wonder if we could solve this via a Price() parameter in offering. 
This gives the cloud provider control without building terse api surface.

I'd lean away from looking up the real cost, but we could simply use a heuristic like (1cpu=1, 1memory=1, 1gpu=1000). We could also use some sort of price threshold, where we don't launch an instance if the threshold isn't met.

Reply:
A single `Price(i InstanceType)` method on cloud provider, or a `Price()` method on the instance provides the ordering the bin packer needs, but doesn't solve the "Don't schedule a non-GPU Workload on a GPU instance type" problem. 

We'll have the non-GPU instance types sorted first in our list provided to the cloud provider `Create()` call.  There's still a chance that with `CreateFleet` and `ec2.SpotAllocationStrategyCapacityOptimizedPrioritized`, you'll get the GPU instance type. 

I suppose the CloudProvider itself could be responsible for further filtering (e.g. if we ask for non-GPU nodes and GPU nodes, the CloudProvider filters out the GPU types).

Reply:
That's where the threshold comes in (e.g. 2-3x). You could design it such that GPUs are simply too expensive.

Reply:
s/weary/wary :)

Agree about the threshold. Perhaps we could make the compare function be able to say "no this is not supported", rather than "prefer this over that".